<a href="https://colab.research.google.com/github/NanjieJA/JC2066-assignment1/blob/main/assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Student Name:

Student ID:

# JC2066 Assignment 1: College Admissions Algorithms

### Learning objectives:
- Explore algorithmic decision-making and representations of people in data
- Understanding outliers, missing information in data, and human judgments in algorithmic decisions

### About Submission
The questions include two parts: **coding** (Q1 to Q5) and **reflection** (Q6 & Q7). After completing all questions, **only upload the finished iPython Notebook file (.ipynb) to Canvas**. Remember to check that you have submitted the file correctly on Canvas (You do not need to submit the files (e.g., csv) you output through your code). **Make sure all cells in your file run without errors after restarting the kernel**.

The deadline to submit your files to Canvas is **Mar. 23rd (Sunday), 11:59 pm**. Please note that this assignment accounts for `20% of your final course score`.

### Background
As you know, the college admissions process involves a lot of types of data from prospective students to make decisions. With the number of applicants increasing, colleges may begin relying on algorithms to select which applications should receive more intensive human review. An algorithm could use quantitative data--such as GPA and SAT score--to provide initial recommendations. In fact, there is more data available than ever. Many colleges even track data about prospective student engagement

> e.g., whether they open emails, visit the college website, engage on social media, etc. This creates a “demonstrated interest” value.

[Based on a survey of college admissions officers](https://www.insidehighered.com/admissions/article/2018/11/12/new-data-admissions-including-application-trends-early-decision-and), we know some of the weights that humans tend to give to these different types of data. Your task will be to create a program that iterates through a list of data points and provides a recommendation for which prospective students are likely to be the best candidates for admission.

Prospective student data is organized in the **admission_algorithms_dataset.csv** file such that the data for each student is on one line, with the values separated by commas. Examples of student data might be:

> Student, SAT, GPA, Interest, High School Quality, Semester 1, Semester 2, Semester 3, Semester 4
>
> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
>
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
>
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

The data includes these columns (in order):

- **Student**: a unique identifier for each student (their name)
- **SAT score**: value between 0 and 1600
- **GPA**: value between 0 and 5
- **Interest**: value between 0 and 10 (from very low interest to very high interest)
- **High School Quality**: value between 0 and 10 (from very low-quality to very high-quality high school)
- **Semester 1**: average grade for semester 1
- **Semester 2**: average grade for semester 2
- **Semester 3**: average grade for semester 3
- **Semester 4**: average grade for semester 4

## **PART 1 CODING TASKS**

### **Question 1 - Getting set up with our data (15 points)**

First, we need to make sure that we can appropriately read in the data line by line, parsing/converting each line into a list and converting each element in the list to the appropriate type.

### Task 1
Read in your data set and loop through its contents line by line. Make use of the `str.split(delimiter)` function to break each individual line into a list of elements (Name, SAT, GPA, etc.). Ensure that you've done this correctly by printing your list after using the split() function. You'll want to delete this print statement later, but make sure to double check that you are correctly looping over all the lines before moving on! Once you have each line in a list, save the student's name in a variable, then delete this element from your list.

In [24]:
from google.colab import files
uploaded = files.upload()

Saving admission_algorithms_dataset.csv to admission_algorithms_dataset (2).csv


In [25]:
filename = "admission_algorithms_dataset.csv"

input_file = open(filename, "r")

# grab the line with the headers
headers = input_file.readline().strip()
### INSERT YOUR CODE HERE

for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]

#for output testing
    print(f"studentName: {student_name}")
    print("otherDatas", line_elements)
input_file.close()


studentName: Abbess Horror 
otherDatas ['1300', '3.61', '10', '7', '95', '86', '91', '94']
studentName: Adele Hawthorne 
otherDatas ['1400', '3.67', '0', '9', '97', '83', '85', '86']
studentName: Adelicia von Krupp 
otherDatas ['900', '4', '5', '2', '88', '92', '83', '72']
studentName: Adriana Soria 
otherDatas ['1360', '3.19', '0', '9', '90', '85', '97', '87']
studentName: Adriana Princess
otherDatas ['1180', '2.93', '2', '7', '93', '81', '90', '87']
studentName: Adrienne Frost 
otherDatas ['1180', '3', '6', '7', '95', '88', '97', '97']
studentName: Agatha Timly 
otherDatas ['1140', '2.98', '10', '8', '92', '81', '95', '90']
studentName: Agent Anesthesia 
otherDatas ['1190', '3.08', '1', '7', '96', '82', '83', '81']
studentName: Aghanne Buffalo
otherDatas ['1400', '3.39', '10', '9', '85', '87', '92', '98']
studentName: Aguja Buffalo
otherDatas ['1310', '3.92', '8', '4', '95', '88', '87', '95']
studentName: Akira Kiamata 
otherDatas ['1310', '4', '1', '1', '90', '86', '81', '82']
stude

### Task 2
Once you have a list of strings for each line, you will write a function `convert_row_type()` that takes one list of elements (representing the data for one student) as a parameter and converts it so that all numbers (currently represented as strings) are converted to floats. Make sure not to lose any information when you do this conversion! Implement this as a pure list function. , i.e., return a new list and do not modify the list passed in.

Example:
> Input: ["1300", "3.61", "10" ,"7" ,"95" ,"86" ,"91" ,"94"]
>
> Return value: [1300.0, 3.61, 10.0, 7.0, 95.0, 86.0, 91.0, 94.0]

A function `check_row_types()` is provided for you to check your results. Once you've called `convert_row_type()` on the list representing one row, you may call this function to ensure that you have correctly converted all elements to floats.

In [26]:
# Provided code
# This function checks to ensure that a list is of length
# 8 and that each element is type float
# Parameters:
# row - a list to check
# Returns True if the length of row is 8 and all elements are floats
def check_row_types(row):
    if len(row) != 8:
        print("Length incorrect! (should be 8): " + str(row))
        return False
    ind = 0
    while ind < len(row):
        if type(row[ind]) != float:
            print("Type of element incorrect: " + str(row[ind]) + " which is " + str(type(row[ind])))
            return False
        ind += 1
    return True

In [17]:
### INSERT YOUR CODE HERE

def convert_row_type(row_list):
    converted = []
    for element in row_list:
        try:
            converted.append(float(element))
        except ValueError:
            converted.append(element)
    return converted

#for testing:#for output testing
def check_row_types(row):
    if len(row) != 8:
        print("Length incorrect! (should be 8): " + str(row))
        return False
    ind = 0
    while ind < len(row):
        if type(row[ind]) != float:
            print("Type of element incorrect: " + str(row[ind]) + " which is " + str(type(row[ind])))
            return False
        ind += 1
    return True
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
headers = input_file.readline().strip()
for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)
    print(check_row_types(converted_data))
input_file.close()

True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


### Task 3
Separate your data. Use list slicing to separate your list (which should contain 8 numbers at this point) into two lists: one that contains the student's SAT, GPA, Interest, and High School Quality scores, and one that contains their 4 semester grades. You'll do Questions 2 - 4 with the first list of 4 numbers and Question 5 with the list of grades.

In [27]:
### INSERT YOUR CODE HERE
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
headers = input_file.readline().strip()
for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)

    student_features = converted_data[:4]
    semester_grades = converted_data[4:]
#for output testing
    print(f"student_features：{student_features}")
    print(f"semester_grades：{semester_grades}")

student_features：[1300.0, 3.61, 10.0, 7.0]
semester_grades：[95.0, 86.0, 91.0, 94.0]
student_features：[1400.0, 3.67, 0.0, 9.0]
semester_grades：[97.0, 83.0, 85.0, 86.0]
student_features：[900.0, 4.0, 5.0, 2.0]
semester_grades：[88.0, 92.0, 83.0, 72.0]
student_features：[1360.0, 3.19, 0.0, 9.0]
semester_grades：[90.0, 85.0, 97.0, 87.0]
student_features：[1180.0, 2.93, 2.0, 7.0]
semester_grades：[93.0, 81.0, 90.0, 87.0]
student_features：[1180.0, 3.0, 6.0, 7.0]
semester_grades：[95.0, 88.0, 97.0, 97.0]
student_features：[1140.0, 2.98, 10.0, 8.0]
semester_grades：[92.0, 81.0, 95.0, 90.0]
student_features：[1190.0, 3.08, 1.0, 7.0]
semester_grades：[96.0, 82.0, 83.0, 81.0]
student_features：[1400.0, 3.39, 10.0, 9.0]
semester_grades：[85.0, 87.0, 92.0, 98.0]
student_features：[1310.0, 3.92, 8.0, 4.0]
semester_grades：[95.0, 88.0, 87.0, 95.0]
student_features：[1310.0, 4.0, 1.0, 1.0]
semester_grades：[90.0, 86.0, 81.0, 82.0]
student_features：[1100.0, 3.22, 10.0, 3.0]
semester_grades：[93.0, 89.0, 94.0, 96.0]
stud

### **Question 2 - Prospective Student Score (15 points)**
### Task 1

Write a function `calculate_score()` that takes a list as a parameter and calculates an overall score for each student based on the following weights:

- 30% SAT
- 40% GPA
- 10% demonstrated interest
- 20% strength of curriculum

The list parameter will contain all of the relevant information about the student. The return value is the student’s calculated score. To make this work, you will also need to normalize both GPA and SAT so that they are also on a 0 to 10 scale. To do this, multiply the GPA by 2, and divide the SAT score by 160.

Example:

> Input: [1300.0, 3.61, 10.0, 7.0] - which represents a student with a 1300 SAT score, a 3.61 GPA, 10 out of 10 for interest and 7 out of 10 for high school quality
>
> Output: ((1300 / 160) * 0.3) + ((3.61 * 2) * 0.4) + (10 * 0.1) + (7 * 0.2) = 7.73 out of 10

Round your final output score to 2 decimal places. This can be done following the example:

> `name = "Dr. Stephens"`
>
> `score = 7.1`
>
> `print(f"{name},{score:.2f}")`

This will print "Dr. Stephens,7.10"

You start the formatted string with an `f`, then enclose the variables in curly braces {}. The `:.2f` tells the formatter to print the variable as a floating point number with 2 digits after the decimal point (`.`). You can use these formatted strings anywhere you would use a string, i.e. print(), write(), & assignment statements, etc.

In [28]:
### INSERT YOUR CODE HERE
def calculate_score(student_features):
  sat, gpa, interest, curriculum = student_features

  sat_normalized = sat / 160
  gpa_normalized = gpa * 2
  weighted_sat = sat_normalized * 0.3
  weighted_gpa = gpa_normalized * 0.4
  weighted_interest = interest * 0.1
  weighted_curriculum = curriculum * 0.2

  total_score = (weighted_sat
          + weighted_gpa
          + weighted_interest
          + weighted_curriculum)

  return round(total_score * 100 + 0.5) / 100
#for output testing
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
headers = input_file.readline().strip()
for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)

    student_features = converted_data[:4]
    feature_score = calculate_score(student_features)

    semester_grades = converted_data[4:]

    print(f"{student_name},{feature_score:.2f}")


Abbess Horror ,7.73
Adele Hawthorne ,7.37
Adelicia von Krupp ,5.79
Adriana Soria ,6.91
Adriana Princess,6.16
Adrienne Frost ,6.62
Agatha Timly ,7.13
Agent Anesthesia ,6.20
Aghanne Buffalo,8.14
Aguja Buffalo,7.20
Akira Kiamata ,5.96
Alaisa Ruatha Pethnan ,6.24
Alchemist Student,7.14
Aldebron James,7.46
Aleksandra Nikolaevna ,6.72
Alflyse Freshman,6.78
Alice Hayes ,6.14
Alice Nugent ,5.73
Alice Tremaine ,6.87
Alizon Baunacha ,7.68
Alkhema Salad,5.52
Allatou Buffalo,6.04
Alma Chalmers ,5.34
Amanda Mueller ,6.30
Amber Alexis ,7.03
Amie Zamborano ,6.73
Amity Hunter ,6.88
Anastasia Kravinoff ,5.81
Anastasia Summit ,7.21
Andrea Haggard ,6.78
Andrea Janson ,7.40
Andrea Roarke ,7.64
Andrea von Strucker ,7.64
Angela Bradford ,7.44
Angela Golden ,7.38
Angelina Brancale ,6.88
Angeline Kutter ,5.53
Anita Ehren ,4.69
Ann Fraley ,6.72
Ann Herd ,5.50
Anna Frankenstein ,5.21
Anne Mutant ,7.55
Anne Marie Cortez ,6.59
Anzhela Federova ,6.65
Aranda Charboneau ,6.62
Araoha Tepania ,7.00
Ariana Saddiqi ,6.1

### Task 2
Call the `calculate_score()` function for each line (row) of data (after you've converted it). Then, write the student's name and their calculated score to a new file called `student_scores.csv` such that each row contains a student’s name and their score, separated by a comma.

Example:

> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
>
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
>
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

lines written to file:

> Abbess Horror , 7.73
>
> Adele Hawthorne , 7.36
>
> Adelicia von Krupp , 5.79

In [29]:
### INSERT YOUR CODE HERE
def calculate_score(student_features):
  sat, gpa, interest, curriculum = student_features

  sat_normalized = sat / 160
  gpa_normalized = gpa * 2
  weighted_sat = sat_normalized * 0.3
  weighted_gpa = gpa_normalized * 0.4
  weighted_interest = interest * 0.1
  weighted_curriculum = curriculum * 0.2

  total_score = (weighted_sat
          + weighted_gpa
          + weighted_interest
          + weighted_curriculum)

  return round(total_score * 100 + 0.5) / 100
#for output testing
filename = "admission_algorithms_dataset.csv"
output_filename = "student_scores.csv"

input_file = open(filename, "r")
output_file = open(output_filename, "w")

headers = input_file.readline().strip()
for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)

    student_features = converted_data[:4]
    feature_score = calculate_score(student_features)

    semester_grades = converted_data[4:]

    output_line = f"{student_name},{feature_score:.2f}\n"
    output_file.write(output_line)

    print(output_line)



Abbess Horror ,7.73

Adele Hawthorne ,7.37

Adelicia von Krupp ,5.79

Adriana Soria ,6.91

Adriana Princess,6.16

Adrienne Frost ,6.62

Agatha Timly ,7.13

Agent Anesthesia ,6.20

Aghanne Buffalo,8.14

Aguja Buffalo,7.20

Akira Kiamata ,5.96

Alaisa Ruatha Pethnan ,6.24

Alchemist Student,7.14

Aldebron James,7.46

Aleksandra Nikolaevna ,6.72

Alflyse Freshman,6.78

Alice Hayes ,6.14

Alice Nugent ,5.73

Alice Tremaine ,6.87

Alizon Baunacha ,7.68

Alkhema Salad,5.52

Allatou Buffalo,6.04

Alma Chalmers ,5.34

Amanda Mueller ,6.30

Amber Alexis ,7.03

Amie Zamborano ,6.73

Amity Hunter ,6.88

Anastasia Kravinoff ,5.81

Anastasia Summit ,7.21

Andrea Haggard ,6.78

Andrea Janson ,7.40

Andrea Roarke ,7.64

Andrea von Strucker ,7.64

Angela Bradford ,7.44

Angela Golden ,7.38

Angelina Brancale ,6.88

Angeline Kutter ,5.53

Anita Ehren ,4.69

Ann Fraley ,6.72

Ann Herd ,5.50

Anna Frankenstein ,5.21

Anne Mutant ,7.55

Anne Marie Cortez ,6.59

Anzhela Federova ,6.65

Aranda Charboneau ,6

Check the `student_scores.csv` file you just create, and print out the answers to the following 3 questions:
- who gets the highest score after calculation, and what is the score?
- who gets the lowest score after calculation, and what is the score?
- what is the average score among all students? Round the results to 2 decimal places.

In [35]:
### INSERT YOUR CODE HERE
import os
def analyze_scores(filename):
    max_score = -float('inf')
    min_score = float('inf')
    total = 0.0
    count = 0
    top_students = []
    bottom_students = []

    if not os.path.exists(filename):
      print(f"Error: file {filename}do not exist.")
      return None

    with open(filename, 'r') as f:
      for line in f:
          line = line.strip()
          if not line:
              continue

          if ',' not in line:
              print(f"Invalid format skipped: {line}")
              continue

          parts = line.split(',', 1)
          if len(parts) != 2:
              print(f"Invalid format skipped: {line}")
              continue

          name, score_str = parts

          if not score_str.replace('.', '', 1).isdigit():
              print(f"Invalid score skipped: {name} - {score_str}")
              continue

          score = float(score_str)

          count += 1
          total += score

          if score > max_score:
              max_score = score
              top_students = [name]
          elif score == max_score:
              top_students.append(name)

          if score < min_score:
              min_score = score
              bottom_students = [name]
          elif score == min_score:
              bottom_students.append(name)

    if count == 0:
        print("Warning: No valid data found")
        return None

    average = round(total / count, 2)

    return {
        "top_students": (top_students, max_score),
        "lowest_students": (bottom_students, min_score),
        "average_score": average
    }

#for output testing
results = analyze_scores("student_scores.csv")

if results:
  print(f"[Highest Score] {', '.join(results['top_students'][0])} : {results['top_students'][1]:.2f}")
  print(f"[Lowest Score] {', '.join(results['lowest_students'][0])} : {results['lowest_students'][1]:.2f}")
  print(f"[Average Score] {results['average_score']:.2f}")
else:
  print("No valid results generated")



[Highest Score] Betty Sumitro  : 8.61
[Lowest Score] Lupe Werewolf : 4.08
[Average Score] 6.65


### Task 3
Write the names for all students who have a score of 6 or higher to a file called `chosen_students.csv`.

Example:

> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
>
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
>
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

lines written to file:

> Abbess Horror
>
> Adele Hawthorne

Make sure to close all the files you opened in your program.


In [36]:
### INSERT YOUR CODE HERE
def choose_students(input_file, output_file):
  with open(input_file, 'r') as infile, open(output_file, 'w') as outfile:

    for line in infile:
        line = line.strip()
        if not line:
            continue

        parts = line.split(',', 1)
        name = parts[0].strip()
        score_str = parts[1].strip()

        score = float(score_str)
        if score >= 6.0:
            outfile.write(f"{name}\n")
#for output testing
choose_students("student_scores.csv", "chosen_students.csv")



Check the `chosen_students.csv` file you just create, how many students are chosen? Print out the number.

In [39]:
### INSERT YOUR CODE HERE
def count(input_file):
  count = 0
  with open(input_file, 'r') as f:
    for line in f:
        cleaned_line = line.strip()
        if cleaned_line:
            count += 1
  return count
#for output testing
print(count("chosen_students.csv"))

159


### **Question 3 - Looking for Outliers (15 points)**
Consider ways that this algorithm might systematically miss certain kinds of edge cases. For example, what if a student has a 0 for demonstrated interest because they don’t use social media or have access to a home computer? What if a student has a very high GPA but their SAT score is low enough to bring their score down; could this mean that they had a single bad test taking day?

### Task 1
Write a function `is_outlier()` that can check for certain kinds of outliers. It should check that:
1. if the demonstrated interest score is 0 or
2. if the normalized GPA that is more than 2 points higher than the normalized SAT score.

If either of these conditions is true, it should return `True` (because this student is an outlier); otherwise, the function returns `False`.

In [42]:
### INSERT YOUR CODE HERE
def is_outlier(featrue_list):
  sat=featrue_list[0]
  gpa=featrue_list[1]
  interest=featrue_list[2]

  return (interest==0) or ((sat/160) - (gpa*2) > 2)



### Task 2
Call `is_outlier()` for each student and write the students' names to a file called `outliers.csv`, one name per line if they are an outlier.

In [43]:
### INSERT YOUR CODE HERE
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
outliers = open("outliers.csv", 'w')
headers = input_file.readline().strip()


for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)

    student_features = converted_data[:4]

    if is_outlier(student_features):
      outliers.write(f"{student_name}\n")






Check the `outliers.csv` file you just create, how many students are identified as outliers? Print out the number.

In [44]:
### INSERT YOUR CODE HERE

print(count("outliers.csv"))

34


### Task 3
Combine the work that you've done now to create an improved list of students to admit to your school. Write students' names, one per line, to the file `chosen_improved.csv` if they either have a score of 6 or greater OR if they are an outlier and their score is 5 or greater. Make sure to take advantage of the work that you’ve already done by calling your functions from previous problems to help you out!

In [48]:
### INSERT YOUR CODE HERE
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
chosen = open("chosen_improved.csv", 'w')
headers = input_file.readline().strip()

for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')

    student_name = line_elements[0]
    del line_elements[0]
    converted_data = convert_row_type(line_elements)
    student_features = converted_data[:4]

    score=calculate_score(student_features)

    con1=is_outlier(student_features) and score >= 5.0
    con2=score >= 6.0

    if con1 or con2 :
      chosen.write(f"{student_name}\n")


Check the `chosen_improved.csv` file you just create, how many students are chosen by this improved algorithm? Print out the number.

In [49]:
### INSERT YOUR CODE HERE

print(count("chosen_improved.csv"))

172


### **Question 4 - Slightly Improved Algorithm (10 points)**
### Task 1

Create a `calculate_score_improved()` function that calculates a student score, checks if it is an outlier, and returns `True` if the student has a score of 6 or higher OR was flagged as an outlier; otherwise, return `False`. Make sure to take advantage of the work that you've already done by calling your functions from previous problems to help you out!

In [51]:
### INSERT YOUR CODE HERE
def calculate_score_improved(student_features):
  score = calculate_score(student_features)
  is_outlier_flag = is_outlier(student_features)

  return (score >= 6.0)or(is_outlier_flag)

### Task 2

Call `calculate_score_improved()` and output each student’s information (name, SAT, GPA, interest score, and high school quality) to a new file called `better_improved.csv` if `calculate_score_improved()` returned `True` for them.

In [53]:
### INSERT YOUR CODE HERE
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
output_file = open("better_improved.csv", 'w')
headers = input_file.readline().strip()

for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')
    name = line_elements[0].strip()
    raw_data = line_elements[1:]
    converted_data = convert_row_type(raw_data)
    student_features = converted_data[:4]

    if calculate_score_improved(student_features):
      output_file.write(f"{name},{student_features[0]},{student_features[1]},{student_features[2]},{student_features[3]}\n")

Check the `better_improved.csv` file you just create, how many students are chosen by this new algorithm? Print out the number.

In [55]:
### INSERT YOUR CODE HERE
print(count("better_improved.csv"))


175


### **Question 5 - GPA Checker (15 points)**

A single GPA score is not a full picture of a student’s academic performance, as it may have improved over time or included outlier courses or semesters. A more context-sensitive algorithm could consider a student’s entire transcript and checks for, for example, a single class score that is more than two letter grades (20 points) lower than all other scores. For this task, you will use the second half of the data for each student in the provided file.

### Task 1
Write a function `grade_outlier()` that takes in a list of grades (of any length) and returns `True` if one single number is more than 20 points lower than all other numbers; otherwise, `False`.

Example:

> Input: [99, 94, 87, 89, 56, 78, 89]

Hint: Sort the list from lowest to highest, and check for the difference between the two lowest grades.

78 - 56 = 22; 22 > 20

> Output: True

Next, consider the data that we have: a list of grades for each student, one grade per semester for four semesters.

Make sure that your `grade_outlier()` function works by calling it for every row in the second dataset. Print out an informative message about which students have a single grade outlier. You'll delete this later but it's a great way of testing your function!

Finally, consider the importance of an algorithm being able to flag students who might have a lower overall GPA but have shown improvement over time.

In [56]:
### INSERT YOUR CODE HERE
def grade_outlier(grades):
  if len(grades) < 2:
      return False

  sorted_grades = sorted(grades)
  if (sorted_grades[1] - sorted_grades[0]) > 20:
    return True
  return False

### Task 2
Create a function `grade_improvement()` that returns `True` if the average score of each semester is higher than or equal to each previous semester and `False` otherwise.

Hint: investigate how the `==` operator works between two lists and think about using the Python's `sorted()` function.

In [58]:
### INSERT YOUR CODE HERE
def grade_improvement(grades):
  sorted_grades = sorted(grades)
  if sorted_grades==grades:
    return True
  return False

### Task 3

Using the grade information that you've just learned, create conditions based on the information from the previous problems and `grade_outlier()` and `grade_improvement()` to choose all students if they either have a score of 6 or greater or if they have a score of 5 or more and at least one of the following is true:

- `is_outlier()` returns `True`
- `grade_outlier()` returns `True`
- or `grade_improvement()` returns `True`

Write the students who fit this description to `composite_chosen.csv`, one name per line.


In [60]:
### INSERT YOUR CODE HERE
filename = "admission_algorithms_dataset.csv"
input_file = open(filename, "r")
output_file = open("composite_chosen.csv", 'w')
headers = input_file.readline().strip()

for line in input_file:
    cleaned_line = line.strip()
    if not cleaned_line:
        continue

    line_elements = cleaned_line.split(',')
    name = line_elements[0].strip()
    raw_data = line_elements[1:]
    converted_data = convert_row_type(raw_data)
    student_features = converted_data[:4]
    studet_grades = converted_data[4:]

    score=calculate_score(student_features)

    condition1=(score>=6.0)
    condition2= (score >= 5.0) and (
      is_outlier(student_features) or
      grade_outlier(studet_grades) or
      grade_improvement(studet_grades)
    )

    if condition1 or condition2:
        output_file.write(f"{name}\n")



Check the `composite_chosen.csv` file you just create, how many students are chosen by this algorithm? Print out the number.

In [62]:
### INSERT YOUR CODE HERE
print(count("composite_chosen.csv"))


175


**Note**:

We will be testing your program using a file with the same format but different data! Make sure that you haven't "hard coded" anything specific to your data.

## **PART 2 REFLECTION TASKS**

Based on the algorithm you just implemented, answer the following questions:

### **Question 6 - Reflect on the outcomes (15 points)**

In 100 to 200 words, reflect on the outcomes of your implemented algorithms. What are some examples of ways these algorithms might still give undesirable results for certain types of prospective students?

Type your answer here:


-

### **Question 7 - How should and shouldn’t algorithms be used in college admissions? Why? (15 points)**

Given the realities of time and resources (i.e., only limited person-hours can be involved in the admission process), do you think we should use algorithms like these for the admission process (assuming the real-world algorithm is far more sophisticated than the one you just created)? Explain your thinking, taking into account possible undesirable outcomes.

Type your answer here:

-